File: app/src/main/java/com/dumuzeyn/mp3player/ui/permissions/DeviceAudioPermissionController.java
package com.dumuzeyn.mp3player.ui.permissions;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;

/** Owns the runtime permission required for automatic MediaStore music discovery. */
public final class DeviceAudioPermissionController {
    public static final int REQUEST_AUDIO_LIBRARY = 34;

    private DeviceAudioPermissionController() {
    }

    public static boolean hasPermission(Context context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return true;
        }
        return context.checkSelfPermission(requiredPermission())
                == PackageManager.PERMISSION_GRANTED;
    }

    public static boolean requestIfNeeded(Activity activity) {
        if (hasPermission(activity)) {
            return false;
        }
        activity.requestPermissions(
                new String[]{requiredPermission()}, REQUEST_AUDIO_LIBRARY);
        return true;
    }

    public static boolean handles(int requestCode) {
        return requestCode == REQUEST_AUDIO_LIBRARY;
    }

    private static String requiredPermission() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU
                ? Manifest.permission.READ_MEDIA_AUDIO
                : Manifest.permission.READ_EXTERNAL_STORAGE;
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/ui/permissions/DeviceAudioPermissionController.kt
package com.dumuzeyn.mp3player.ui.permissions

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build

/** Owns the runtime permission required for automatic MediaStore music discovery. */
object DeviceAudioPermissionController {
    const val REQUEST_AUDIO_LIBRARY = 34

    @JvmStatic
    fun hasPermission(context: Context): Boolean =
        Build.VERSION.SDK_INT < Build.VERSION_CODES.M ||
            context.checkSelfPermission(requiredPermission()) == PackageManager.PERMISSION_GRANTED

    @JvmStatic
    fun requestIfNeeded(activity: Activity): Boolean {
        if (hasPermission(activity)) return false
        activity.requestPermissions(arrayOf(requiredPermission()), REQUEST_AUDIO_LIBRARY)
        return true
    }

    @JvmStatic
    fun handles(requestCode: Int): Boolean = requestCode == REQUEST_AUDIO_LIBRARY

    private fun requiredPermission(): String =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            Manifest.permission.READ_MEDIA_AUDIO
        } else {
            Manifest.permission.READ_EXTERNAL_STORAGE
        }
}
